### PR TITLE
bug-1692-use-tradebar-data-when-no-quotebar-data-is-available

### PR DESCRIPTION
**Pull request by Jay-Jay-D**
#### Description

This PR implements the following chages in the `SecurityCache` class:
- fill-forward data is not cached.
- in case of multiple `MarketDataType`it returns the one with the latest timestamp.
- prioritize the quotebars.

#### Related Issue

#1692 

#### Motivation and Context

Before this PR, if no quote data was available, then the algorithm equity  was flat until new quote data was available.

#### How Has This Been Tested?

The method for testing was using an basic template algorithm (it buys 100% of one security at the beginning) and confirm the equity follows perfectly the security close data when there is only tradebar data available. 

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [ ] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by Jay-Jay-D** [comment]
The `BasicTemplateCryptoAlgorithm` sends a trade at each hour (buy or liquidate). And all have the same value, except for the filling price of the Order #21 fired at 2016-10-07 20:00 algorithm time, equivalent to 2016-10-08 00:00 UTC. 
In order to fill that order:
- The engine uses the correspondent price, the 2016-10-07 23:59 UTC minute price.
- The last time stamp of the file `20161007_quote.zip` is 86280000, equivalent to 2016-10-07 *23:58* UTC minute price
- With the actual `SecurityCache` behavior, the 2016-10-07 23:58 UTC minute price (618.33) is fill forward as *quote* and the engine uses that price for filling the order #21.
- With the suggested change in the `SecurityCache`, the engine uses the 2016-10-07 23:59 UTC *tradebar* price (617.98) from the `20161007_trade.zip`.

That small difference explains the changes in the regression backtest statistics.

File: Common/Securities/SecurityCache.cs [hunk 1]
```diff
@@ -101,10 +101,27 @@ public void AddData(BaseData data)
...
             }
 
-            _lastData = data;
-            _dataByType[data.GetType()] = data;
+            // Only cache no fill-forward data.
+            if (data.IsFillForward) return;
 
-            var tick = data as Tick;
+            // If the _dataByType dictionary is not populated or it only has one MarketDataType, just use the last value.
+            if (_dataByType.Keys.Count <= 1)
```
Comment (jaredbroad) [inline Common/Securities/SecurityCache.cs]: Will need to refactor this monday but will merge for now.

File: Common/Securities/SecurityCache.cs [hunk 2]
```diff
@@ -101,10 +101,27 @@ public void AddData(BaseData data)
...
+            {
+                _lastData = data;
+                _dataByType[data.GetType()] = data;
+            }
+            else
+            {
+                // If there are more then one MarketDataType then update _lastData only if is the latest, prioritizing QuoteBars
+                if (data.Time > _lastData.Time || data.Time == _lastData.Time && data.DataType == MarketDataType.QuoteBar)
```
Comment (mchandschuh) [inline Common/Securities/SecurityCache.cs]: We need to check on the `EndTime` property.

File: Tests/RegressionTests.cs [hunk 3]
```diff
@@ -705,19 +705,19 @@ private static TestCaseData[] GetRegressionTestParameters()
...
                 {"Profit-Loss Ratio", "0"},
                 {"Alpha", "-11.165"},
-                {"Beta", "574.85"},
-                {"Annual Standard Deviation", "0.353"},
-                {"Annual Variance", "0.125"},
-                {"Information Ratio", "-20.427"},
-                {"Tracking Error", "0.353"},
-                {"Treynor Ratio", "-0.013"},
-                {"Total Fees", "$6076.08"}
+                {"Beta", "585.081"},
```
Comment (mchandschuh) [inline Tests/RegressionTests.cs]: I'd be curious how these specifically changed. For example, were we previously filling using fill-forward trade data? Or perhaps we were filling using trade data and now it's filling using quote data? Changes in regression can be expected with bugfixes but I think it's critically important to be able to completely explain *why* the expected results have changed.

**Pull request merged**